File: consumer.py

from queue import Queue
from threading import Event
from py_client.message.message import Message
from py_client.message.message_type import MessageType


class Consumer:

    def __init__(self, host: str, port: int, consumer_id: str):
        from py_client.client import Client
        self.client = Client(host, port, consumer_id)

    def __enter__(self):
      return self

    def __exit__(self, exc_type, exc_value, exc_traceback):
        self.client.close()

    def consume(self, message: Message):
        message.type = MessageType.REQ_PULL
        return self.client.fetch(message).result()
    
    def find(self, message: Message):
        message.type = MessageType.REQ_FIND
        response = self.client.fetch(message).result()
        return int(response.get_header("offset", "-1"))
    
    def seek(self, message: Message):
        message.type = MessageType.REQ_SEEK
        response = self.client.fetch(message).result()
        return response.get_header("ok", "false").lower() == "true"

File: message_broker/app/src/main/java/capstone/design/client/Consumer.java
package capstone.design.client;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import capstone.design.message.Message;
import capstone.design.message.MessageType;
import capstone.design.netty.client.NettyClient;


public class Consumer implements AutoCloseable {
    
    private final NettyClient client;

    // constructor =====
    public Consumer(String host, int port, String clientId) throws Exception { 
        this.client = new NettyClient(host, port, clientId);
    }

    // method =====
    public List<Message> consume(Message message) {
        message.setType(MessageType.REQ_PULL);
        List<Message> responses = new ArrayList<>();
        for (CompletableFuture<Message> future : client.fetch(message)) {
            try {
                responses.add(future.join());
            } catch (Exception e) {
                System.err.println("? Consumer.consume(): future 예외 발생: " + e);
            }
        }

        return responses;
    }

    public int find(Message message) {
        message.setType(MessageType.REQ_FIND);
        try {
            Message response = client.fetch(message).get(0).join();
            return Integer.parseInt(response.header("offset", "-1"));
        } catch (Exception e) {
            System.err.println("? Consumer.find(): " + e);
            return -1;
        }
    }

    public boolean seek(Message message) {
        message.setType(MessageType.REQ_SEEK);
        try {
            Message response = client.fetch(message).get(0).join();
            return response.header("error", "").isEmpty();
        } catch (Exception e) {
            System.err.println("? Consumer.seek(): " + e);
            return false;
        }
    }

    @Override
    public void close() throws Exception { 
        client.shutdown();
    }
}
